matrix/README.md

```diff
@@ -0,0 +1,300 @@
+    for d in [None, int, str]:
+        yield d
+
+@matrix(
+    a=gen,                        # generator function
+    b=(b for b in (True, False)), # generator
+    c=42,                         # fixed value
+    d=range(5),                   # iterable object (via __iter__)
+    e=["odd", "even"],            # sequence (set, tuple, list)
+)
+def foo(a, b, c, d):
+    pass
+```
+
+See the note below in [Dynamic parameters](#generator-functions) regarding use of
+generator functions.
+
+## Return values
+
+The return value from calling the decorated function will be a sequence of all
+the return values for each invocation of the function with parameters from rows
+of the matrix. If the function raises an exception, the caught exception will be
+the return value for that call.
+
+```python
+@matrix(a=range(1, 3), b=range(3))
+def division(a, b):
+    return a / b
+
+print(division())
+```
+
+Will produce the following output:
+
+```
+[
+    ZeroDivisionError('division by zero'), # 1 / 0
+    1.0,                                   # 1 / 1
+    0.5,                                   # 1 / 2
+    ZeroDivisionError('division by zero'), # 2 / 0
+    2.0,                                   # 2 / 1
+    1.0,                                   # 2 / 2
+]
+```
+
+## Partial parameter specification
+
+It is also possible to only specify a subset of a function's arguments as matrix
+parameters. Any arguments not defined with the `@matrix` decorator must be
+specified when the function is invoked.
+
+```python
+@matrix(x=range(1, 4))
+def quadratic(a, x, b):
+    return (a * x) + b
+
+assert quadratic(a=2, b=10) == [12, 14, 16]
+```
+
+## Post-definition overrides
+
+Parameters may be overridden after definition:
+
+```python
+print_xy.override(x=[7])()
+```
+
+```
+x=7, y=1.5
+x=7, y=6.5
+```
+
+## Dynamic parameters
+
+Dynamic or dependent parameters may be generated using a callable rather than a
+value range.
+
+```python
+def y(x):
+    if x == 1:
+        return [1.5, 6.5]
+    return [1.7, 2.3, 3.4]
+
+@matrix(
+    x=range(1, 3),
+    y=y,
+)
+def print_xy(x, y):
+    print(f"{x=}, {y=}")
+
+print_xy()
+```
+
+```
+x=1, y=1.5
+x=1, y=6.5
+x=2, y=1.7
+x=2, y=2.3
+x=2, y=3.4
+```
+
+This is roughly equivalent to the following:
+
+```python
+def y(x):
+    if x == 1:
+        return [1.5, 6.5]
+    return [1.7, 2.3, 3.4]
+
+def print_xy(x, y):
+    print(f"{x=}, {y=}")
+
+for x in range(1, 3):
+    for _y in y(x=x):
+        print_xy(x=x, y=_y)
+```
+
+The callable will be invoked with all prior generated parameters, in the order
+they are defined in the `@matrix` decorator. It is possible to specify multiple
+dynamic parameters, where the latter parameters depend on the former.
+
+### Generator Functions
+
+A custom generator function as a parameter range is effectively equivalent to a
+dynamic parameter range. This is intentional, since generators cannot be rewound
+and it is often necessary to iterate through the range of values multiple times.
```

Comment:
and it is often necessary to iterate multiple times through a `matrix` parameter space.